1-static-analysis.md

```diff
@@ -0,0 +1,298 @@
+any of its transitive dependencies change, invaliding any intermediate modules,
+and allowing for graceful hand-off of module scoped stage when necessary.
+
+<details>
+  <summary>Hot module replacement sketch</summary>
+
+  This sketch outlines how one can use `Module` and `ModuleSource` to construct
+  a watcher graph that reuses these objects between reloads when possible.
+  The sketch assumes the existence of a fictitious `watch` interface that is a
+  parody of `fetch`, except producing a promise `changed` that will settle when
+  the response is no longer valid.
+
+  ```js
+  const getImports = source => source.bindings.map(binding =>
+    binding.from ??
+    binding.importAllFrom ??
+    binding.exportAllFrom
+  ).filter(Boolean);
+
+  const sources = new Map();
+  const modules = new Map();
+  const watchers = new Map();
+  const states = new Map();
+  const getStates = new Map();
+
+  const invalidateModule = url => {
+    const watcher = watchers.get(url);
+    if (watcher) {
+      watcher();
+      watchers.delete(url);
+    }
+    modules.delete(url);
+    for (const importSpecifier of getImports(source)) {
+      const url = new URL(importSpecifier, url).href;
+      invalidateModule(url);
+    }
+
+    // Hand-off state in preparation for an upgrade.
+    const getState = getStates.get(url);
+    if (getState) {
+      states.set(url, getState());
+      getStates.delete(url);
+    }
+  };
+
+  const invalidateSource = url => {
+    invalidateModule(url);
+    sources.delete(url);
+  };
+
+  const importHook = async (importSpecifier, importerMeta) => {
+    const url = new URL(importSpecifier, importerMeta.url).href;
+    let module = modules.get(url);
+    if (!module) {
+      let source = sources.get(url);
+      if (!source) {
+        const response = await watch(url);
+        response.changed.then(() => invalidateSource(url));
+        const text = await response.text();
+        source = new ModuleSource(text);
+        sources.set(url, source);
+      }
+      const registerGetState = getState => {
+        getStates.set(url, getState);
+      };
+      const state = stages.get(url);
+      const importMeta = { url, state, registerGetState };
+      module = new Module(source, { importHook, importMeta });
+      modules.set(url, module);
+    }
+    return module;
+  }
+
+  const watchModule = async (url, { signal }) => {
+    while (!signal.aborted) {
+      const { promise, resolve } = Promise.defer();
+      watchers.set(url, resolve);
+      await importHook(url, import.meta);
+      await promise;
+      // Blink once to debounce coincident changes.
+      await Promise.delay(100);
+    }
+  };
+
+  const entrypoint = 'https://example.com/example.js';
+  await watchModule(entrypoint);
+  ```
+
+  This assumes a protocol for state hand-off:
+
+  ```js
+  let state = import.meta.state;
+  import.meta.registerGetState(() => state);
+  ```
+
+</details>
+
+## Design Questions
+
+Do we also need to reflect `isAsync`?
```

Comment:
this is again, another internal meta data of the source that should probably not be exposed.

Reply:
Yes, I leave this open because revealing it might be necessary for @Jack-Works use case.

Reply:
This is currently required to be known before execution in the spec algorithm of Execute() method of a Cyclic Module Record.

I'm currently implementing this proposal following the Cyclic Module Record's algorithm. If we have a new spec for it that does not need to know this information, it is not necessary to me.

But I don't agree it's "internal" metadata. This reflects if the module is an Async module or not.